signer/src/stacks_api.rs

```diff
@@ -40,154 +36,324 @@ impl StacksClient {
-        let path = format!("/extended/v2/burn-blocks/0x{}", hash);
-        let url = self.api_endpoint.join(&path).map_err(Error::PathParse)?;
-
-        tracing::debug!(%hash, "Fetching block IDs confirmed by bitcoin block from stacks API");
+    /// Fetch the raw stacks nakamoto block from a Stacks node given the
+    /// Stacks block ID.
+    async fn get_block(&self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
+        let path = format!("/v3/blocks/{}", block_id.to_hex());
+        let base = self.node_endpoint.clone();
+        let url = base
+            .join(&path)
+            .map_err(|err| Error::PathJoin(err, base, Cow::Owned(path)))?;
+
+        tracing::debug!(%block_id, "Making request to the stacks node for the raw nakamoto block");
 
         let response = self
             .client
-            .get(url.clone())
+            .get(url)
             .timeout(REQUEST_TIMEOUT)
             .send()
             .await
-            .map_err(|err| Error::StacksNodeRequest(err, url.clone()))?;
-        let resp: GetBurnBlockResponse = response
-            .json()
+            .map_err(Error::StacksNodeRequest)?;
+        let resp = response
+            .bytes()
             .await
-            .map_err(|err| Error::UnexpectedStacksResponse(err, url))?;
-
-        // The Stacks API often returns hex prefixed with 0x. If this is
-        // the case, we split it off before constructing the block ids.
-        resp.stacks_blocks
-            .into_iter()
-            .map(|hex_string| {
-                let hex_str: &str = if hex_string.starts_with("0x") {
-                    hex_string.split_at(2).1
-                } else {
-                    &hex_string
-                };
-
-                StacksBlockId::from_hex(hex_str)
-                    .map_err(|err| Error::ParseStacksBlockId(err, hex_string))
-            })
-            .collect()
+            .map_err(Error::UnexpectedStacksResponse)?;
+
+        NakamotoBlock::consensus_deserialize(&mut &*resp)
+            .map_err(|err| Error::DecodeNakamotoBlock(err, block_id))
     }
 
-    /// Fetch the raw stacks nakamoto block from a Stacks node given the
-    /// Stacks block ID.
-    async fn get_block(&self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
-        let path = format!("/v3/blocks/{}", block_id.to_hex());
-        let url = self.node_endpoint.join(&path).map_err(Error::PathParse)?;
+    /// Fetch all Nakamoto ancestor blocks within the same tenure as the
+    /// given block ID from a Stacks node.
+    ///
+    /// The response includes the Nakamoto block for the given block id.
+    async fn get_blocks(&self, block_id: StacksBlockId) -> Result<Vec<NakamotoBlock>, Error> {
+        let mut blocks = Vec::new();
+
+        tracing::debug!(%block_id, "Making initial request for Nakamoto blocks within the tenure");
+        blocks.extend(self.get_tenure(block_id).await?);
```

Comment:
```suggestion
    /// The response includes the Nakamoto block for the given block id.
    [tracing::instrument(skip(self))]
    async fn get_blocks(&self, block_id: StacksBlockId) -> Result<Vec<NakamotoBlock>, Error> {
        let mut blocks = Vec::new();

        tracing::debug!("Making initial request for Nakamoto blocks within the tenure");
        blocks.extend(self.get_tenure(block_id).await?);
```

Reply:
Same here, we can instrument the function call and skip the explicit block ID parameter to the log call,